### PR TITLE
Create LDAP server and complete login endpoint - conflicts fixed

### PR DESCRIPTION
**Pull request by aiyan**
This implements PLACES-132:
https://pathcheck.atlassian.net/browse/PLACES-132

Notes:
Since SPL isn't managing the user identities, this is merely a stab at premature integration.
This relies on the assumption that the HA is using Lightweight Directory Access Protocol (LDAP) for identity management and RBAC for permissioning.

The LDAP server (labeled ldapjs) in the backend is for testing purposes only. It is not meant to be used for production.

The HA integration team will likely adjust the authentication middleware to connect to their server or use another scheme altogether.

Closes #71 - it is the same code modifications, except the merge conflicts are fixed.

### CODE REVIEW

**Comment by aiyan** [comment]
@sherifkozman The build issue may be due to incorrect environment variables.

**Comment by sherifkozman** [comment]
so CI works fine however they dont seem to pass the tests against API tests by @AdamLeonSmith 
You should run this with @AdamLeonSmith and @sublet

**Comment by aiyan** [comment]
Gonna take a stab here and say the login test isn't passing. @AdamLeonSmith I changed the credentials from `admin:admin` to `admin:password`.

**Comment by AdamLeonSmith** [comment]
I'm afraid this looks like a docker/CI issue @sherifkozman - at least it does on the latest run!

File: ldapjs/README.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# LDAP Server
+
+This basic LDAP server is to be used for testing purposes only.
+It is not meant for production use.
+
+Please see the OpenLDAP implementations for production-ready servers.
```
Comment (krishnadurai) [inline ldapjs/README.md:6]: Why shouldn't we use OpenLDAP for testing? They do have a docker image.

File: ldapjs/index.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
+const ldap = require('ldapjs');
+const fs = require('fs');
+const path = require('path');
+
+const server = ldap.createServer();
+
+server.bind('cn=root', function(req, res, next) {
```
Comment (krishnadurai) [inline ldapjs/index.js:7]: Would it be better to have this configurable rather than a string here?

File: src/server/passport.js [hunk 3]
```diff
@@ -71,11 +33,35 @@ const jwtStrategy = new JWTstrategy(opts, async (jwt_payload, done) => {
 
 passport.use('jwt', jwtStrategy);
 
-passport.serializeUser(function (user, done) {
+const ldapClient = ldap.createClient({
+  url: 'ldap://127.0.0.1:1389',
```
Comment (krishnadurai) [inline src/server/passport.js:37]: URL should be an ENV variable.

**Comment by aiyan** [comment]
> @aiyan Before I review further, are we having a consensus that LDAP will be used at the mode of login for SPL?
> 
> My suggestion is to have authentication support OIDC protocol which would accomodate a variety of authentication mechanisms.

So I didn't bother getting support for OIDC because I figured it was likely that it may happen in the API proxy layer and not in the application middleware.

File: ldapjs/README.md (hunk 1 above)
Comment (aiyan) [reply ldapjs/README.md:6]: To be honest, I think using OpenLDAP may be overkill, so I spun up a lightweight NodeJS implementation of an LDAP server. Both are essentially equivalent, and the HA is going to fine-tune the passport settings anyway.

File: ldapjs/index.js (hunk 2 above)
Comment (aiyan) [reply ldapjs/index.js:7]: Yes, I will change this.

File: src/server/passport.js (hunk 3 above)
Comment (aiyan) [reply src/server/passport.js:37]: Sure.

**Comment by aiyan** [comment]
Ok I patched all issues raised by @krishnadurai and made sure the tests work. This should be ready to merge.

P.S. The server now requires `/ldapjs/index.js` to be run for Passport functionality.
Moreover, 2 environment variables were added. See `/.env.template` for more info.

**Comment by aiyan** [comment]
@sherifkozman I see this docker error in the build step:
```
Run docker build -t "$REGISTRY_HOSTNAME"/"$GKE_PROJECT"/"$IMAGE":"$GITHUB_SHA" \
18
invalid argument "gcr.io//safeplaces-be-express-test:e1a87bdf389e0e9fbe1df98ee9db41dfe091f732" for "-t, --tag" flag: invalid reference format
```
Is this the new code or is this something to do with CI?

File: ldapjs/ldapServer.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    }
+  });
+
+  return next(new Error('Not found'));
+});
+
+module.exports = {
+  start: () => {
+    return new Promise(resolve => {
+      server.listen(1389, function () {
```
Comment (krishnadurai) [inline ldapjs/ldapServer.js:68]: Read from ENV

Comment (aiyan) [reply ldapjs/ldapServer.js:68]: Will do!

**Comment by aiyan** [comment]
@krishnadurai All issues are fixed.

**Pull request closed without merging**